src/lib/ffi/ffi.cp and "error: ‘q’ was not declared in this scope"

This surfaced under [Travis-CI](https://travis-ci.org/randombit/botan/jobs/222578983):

```
ccache g++ -fstack-protector -m64 -pthread -fPIC -fvisibility=hidden -std=c++11 -D_REENTRANT -g -O3 -momit-leaf-frame-pointer -Wall -Wextra -Wpedantic -Wstrict-aliasing -Wstrict-overflow=5 -Wcast-align -Wmissing-declarations -Wpointer-arith -Wcast-qual -Wzero-as-null-pointer-constant -Wnon-virtual-dtor -Ibuild/include -c ./src/lib/mac/poly1305/poly1305.cpp -o build/obj/lib/mac_poly1305.o

./src/lib/ffi/ffi.cpp: In function ‘int botan_pubkey_load_elgamal(botan_pubkey_struct**, botan_mp_t, botan_mp_t, botan_mp_t)’:

./src/lib/ffi/ffi.cpp:1450:22: error: ‘q’ was not declared in this scope

    BOTAN_UNUSED(q);

                      ^

./src/lib/ffi/ffi.cpp:1452:22: error: ‘x’ was not declared in this scope

    BOTAN_UNUSED(x);

                      ^
```

--- a/src/lib/ffi/ffi.cpp
+++ b/src/lib/ffi/ffi.cpp
@@ -1438,10 +1438,7 @@ int botan_pubkey_load_elgamal(botan_pubkey_t* key,
 
    return -1;
 #else
-   BOTAN_UNUSED(p);
-   BOTAN_UNUSED(q);
-   BOTAN_UNUSED(g);
-   BOTAN_UNUSED(x);
+   BOTAN_UNUSED(key, p, g, y);
    return BOTAN_FFI_ERROR_NOT_IMPLEMENTED;
 #endif
    }
@@ -1464,9 +1461,7 @@ int botan_privkey_load_elgamal(botan_privkey_t* key,
       }
    return -1;
 #else
-   BOTAN_UNUSED(p);
-   BOTAN_UNUSED(g);
-   BOTAN_UNUSED(x);
+   BOTAN_UNUSED(key, p, g, x);
    return BOTAN_FFI_ERROR_NOT_IMPLEMENTED;
 #endif
    }

--- a/src/lib/ffi/ffi.cpp
+++ b/src/lib/ffi/ffi.cpp
@@ -1332,9 +1332,8 @@ int botan_pubkey_load(botan_pubkey_t* key,
 int botan_privkey_load_rsa(botan_privkey_t* key,
                            botan_mp_t p, botan_mp_t q, botan_mp_t d)
    {
-   *key = nullptr;
-
 #if defined(BOTAN_HAS_RSA)
+   *key = nullptr;
    try
       {
       *key = new botan_privkey_struct(new Botan::RSA_PrivateKey(safe_get(p),
@@ -1348,6 +1347,7 @@ int botan_privkey_load_rsa(botan_privkey_t* key,
       }
    return -1;
 #else
+   BOTAN_UNUSED(key, p, q, d);
    return BOTAN_FFI_ERROR_NOT_IMPLEMENTED;
 #endif
    }
@@ -1355,9 +1355,8 @@ int botan_privkey_load_rsa(botan_privkey_t* key,
 int botan_pubkey_load_rsa(botan_pubkey_t* key,
                           botan_mp_t n, botan_mp_t e)
    {
-   *key = nullptr;
-
 #if defined(BOTAN_HAS_RSA)
+   *key = nullptr;
    try
       {
       *key = new botan_pubkey_struct(new Botan::RSA_PublicKey(safe_get(n), safe_get(e)));
@@ -1370,6 +1369,7 @@ int botan_pubkey_load_rsa(botan_pubkey_t* key,
 
    return -1;
 #else
+   BOTAN_UNUSED(key, n, e);
    return BOTAN_FFI_ERROR_NOT_IMPLEMENTED;
 #endif
    }
@@ -1377,9 +1377,8 @@ int botan_pubkey_load_rsa(botan_pubkey_t* key,
 int botan_privkey_load_dsa(botan_privkey_t* key,
                            botan_mp_t p, botan_mp_t q, botan_mp_t g, botan_mp_t x)
    {
-   *key = nullptr;
-
 #if defined(BOTAN_HAS_DSA)
+   *key = nullptr;
    try
       {
       Botan::Null_RNG null_rng;
@@ -1393,10 +1392,7 @@ int botan_privkey_load_dsa(botan_privkey_t* key,
       }
    return -1;
 #else
-   BOTAN_UNUSED(p);
-   BOTAN_UNUSED(q);
-   BOTAN_UNUSED(g);
-   BOTAN_UNUSED(x);
+   BOTAN_UNUSED(key, p, q, g, x);
    return BOTAN_FFI_ERROR_NOT_IMPLEMENTED;
 #endif
    }
@@ -1404,9 +1400,8 @@ int botan_privkey_load_dsa(botan_privkey_t* key,
 int botan_pubkey_load_dsa(botan_pubkey_t* key,
                           botan_mp_t p, botan_mp_t q, botan_mp_t g, botan_mp_t y)
    {
-   *key = nullptr;
-
 #if defined(BOTAN_HAS_DSA)
+   *key = nullptr;
    try
       {
       Botan::DL_Group group(safe_get(p), safe_get(q), safe_get(g));
@@ -1420,10 +1415,7 @@ int botan_pubkey_load_dsa(botan_pubkey_t* key,
 
    return -1;
 #else
-   BOTAN_UNUSED(p);
-   BOTAN_UNUSED(q);
-   BOTAN_UNUSED(g);
-   BOTAN_UNUSED(y);
+   BOTAN_UNUSED(key, p, q, g, y);
    return BOTAN_FFI_ERROR_NOT_IMPLEMENTED;
 #endif
    }

--- a/src/lib/tls/tls_callbacks.h
+++ b/src/lib/tls/tls_callbacks.h
@@ -188,9 +188,7 @@ class BOTAN_DLL Callbacks
        */
        virtual void tls_log_debug_bin(const char* descr, const uint8_t val[], size_t val_len)
           {
-          BOTAN_UNUSED(descr);
-          BOTAN_UNUSED(val);
-          BOTAN_UNUSED(val_len);
+          BOTAN_UNUSED(descr, val, val_len);
           }
    };
 

--- a/src/lib/utils/assert.h
+++ b/src/lib/utils/assert.h
@@ -1,6 +1,7 @@
 /*
 * Runtime assertion checking
 * (C) 2010 Jack Lloyd
+*     2017 Simon Warta (Kullo GmbH)
 *
 * Botan is released under the Simplified BSD License (see license.txt)
 */
@@ -87,9 +88,32 @@ BOTAN_NORETURN void BOTAN_DLL assertion_failure(const char* expr_str,
    } while(0)
 
 /**
-* Mark variable as unused
+* Mark variable as unused. Takes between 1 and 9 arguments and marks all as unused,
+* e.g. BOTAN_UNUSED(a); or BOTAN_UNUSED(x, y, z);
 */
-#define BOTAN_UNUSED(v) static_cast<void>(v)
+#define _BOTAN_UNUSED_IMPL1(a)                         static_cast<void>(a)
+#define _BOTAN_UNUSED_IMPL2(a, b)                      static_cast<void>(a); _BOTAN_UNUSED_IMPL1(b)
+#define _BOTAN_UNUSED_IMPL3(a, b, c)                   static_cast<void>(a); _BOTAN_UNUSED_IMPL2(b, c)
+#define _BOTAN_UNUSED_IMPL4(a, b, c, d)                static_cast<void>(a); _BOTAN_UNUSED_IMPL3(b, c, d)
+#define _BOTAN_UNUSED_IMPL5(a, b, c, d, e)             static_cast<void>(a); _BOTAN_UNUSED_IMPL4(b, c, d, e)
+#define _BOTAN_UNUSED_IMPL6(a, b, c, d, e, f)          static_cast<void>(a); _BOTAN_UNUSED_IMPL5(b, c, d, e, f)
+#define _BOTAN_UNUSED_IMPL7(a, b, c, d, e, f, g)       static_cast<void>(a); _BOTAN_UNUSED_IMPL6(b, c, d, e, f, g)
+#define _BOTAN_UNUSED_IMPL8(a, b, c, d, e, f, g, h)    static_cast<void>(a); _BOTAN_UNUSED_IMPL7(b, c, d, e, f, g, h)
+#define _BOTAN_UNUSED_IMPL9(a, b, c, d, e, f, g, h, i) static_cast<void>(a); _BOTAN_UNUSED_IMPL8(b, c, d, e, f, g, h, i)
+#define _BOTAN_UNUSED_GET_IMPL(_1, _2, _3, _4, _5, _6, _7, _8, _9, IMPL_NAME, ...) IMPL_NAME
+
+#define BOTAN_UNUSED(...) _BOTAN_UNUSED_GET_IMPL(__VA_ARGS__,                      \
+                                                 _BOTAN_UNUSED_IMPL9,              \
+                                                 _BOTAN_UNUSED_IMPL8,              \
+                                                 _BOTAN_UNUSED_IMPL7,              \
+                                                 _BOTAN_UNUSED_IMPL6,              \
+                                                 _BOTAN_UNUSED_IMPL5,              \
+                                                 _BOTAN_UNUSED_IMPL4,              \
+                                                 _BOTAN_UNUSED_IMPL3,              \
+                                                 _BOTAN_UNUSED_IMPL2,              \
+                                                 _BOTAN_UNUSED_IMPL1,              \
+                                                 unused dummy rest value           \
+                          ) /* we got an one of _BOTAN_UNUSED_IMPL*, now call it */ (__VA_ARGS__)
 
 }
 

